### ISSUE TITLE
zero extension tends to generate ugly code

### ISSUE THREAD

**Issue by jrmuizel** [yegord/snowman#34]
It seems relatively common for compilers to take advantage of the implicit zero extension.
Currently snowman will split:
```
xorl	%edx, %edx
```
into:
```c
        *(int32_t*)&rdx4 = 0;
        *((int32_t*)&rdx4 + 1) = 0;
```
An example binary can be found here: http://people.mozilla.org/~jmuizelaar/snowman/f.dylib
The source is the same as issue #30

**Comment by nihilus** [comment]
Did you see my MIPS 'multi' generated-code? Which for some reason like yours ends up being 'signed' and not unsigned. :-/

**Comment by hlide** [comment]
yeah, this is a common optimization for using `xor %edx, %edx` to save one byte instead of `xor %rdx, %rdx` for the intent of having `%rdx = 0`.

**Comment by hlide** [comment]
instead of having:
```c++
*(int32_t*)&rdx4 = expr;
*((int32_t*)&rdx4 + 1) = 0;
```
it is better to do this way:
```c++
rdx4 = (int64_t)expr;
```
if *expr* is signed

**or**

```c++
rdx4 = (uint64_t)expr;
```
if *expr* is unsigned

that is, if the assigned term is a register, it must be preserved and the right term to be promoted to the same size of left term instead of splitting the left term into two.

**Comment by nihilus** [comment]
Geesh, remember me of this stupid manager at QBranch AB whom told me / bragged about about when he asked some what you use 'xor' for... And proudly told me had got this guy and told him that you use it to nullify a register... When I told him that you can do that but it is an exclusive or between two arbitary registers he was really about to bite some dust. ;-)

**Comment by hlide** [comment]
One solution may be to let *operand(index)* return a template class which derives of the expected class of the old *operand(index)* so the operator **^=** may be overloaded to promote a 32-bit register under *operand(index)* into the 64-bit one and add a zero-extend to the right term of the operator **^=**. If *operand(index)* is not used as a left term of operator **^=**, it should be implicitly cast-able into a term that operator like **+** can handle.

**Comment by yegord** [comment]
On Mon, Aug 03, 2015 at 08:21:50AM -0700, Jeff Muizelaar wrote:
> *(int32_t*)&rdx4 = 0;
> *((int32_t*)&rdx4 + 1) = 0;

Yes, this should be fixed. Patches are welcome. I thought about the
problem for a while, and still do not have good solutions.

-- 
Yegor Derevenets
